Fix deadlock in login handler

thallgren opened the pull request:
## Description

The access-token "refresh" routine has a select case that calls the
`loginExecutor.getToken` method which in turn would retrieve the `Token`
and potentially end up in a callback writing to the `refreshTimerReset`
channel that was read by another case in that same select, which lead
to a dead-lock because there would be nothing reading that channel.

This commit ensures that the `loginExecutor.getToken` method is called
in a separate goroutine to avoid blocking the for-loop containing the
select.

## Checklist

<!--
  Please review the requirements for each checkbox, and check them
  off (change "[ ]" to "[x]") as you verify that they are complete.
-->

 - [x] I made sure to update `./CHANGELOG.md`.
 - [x] My change is adequately tested.

LukeShu commented on CHANGELOG.md:
```diff
@@ -31,6 +31,9 @@
 - Change: Failure to report metrics is logged using loglevel info rather than error.
 
+- Bugfix: A potential deadlock situation is fixed that sometimes caused the user daemon to become a zombie process when the user
+  was logged in. 
```
I normally understand the term "zombie process" in the unix sense: "a process that has exited but whose parent has not yet `wait()ed` on it", which I don't think is what's going on here.

LukeShu commented on pkg/client/connector/userd_auth/login.go:
```diff
@@ -233,12 +233,17 @@ func (l *loginExecutor) Worker(ctx context.Context) error {
-				}
+				// Call getToken() in separate goroutine because it may end up
+				// writing to the refreshTimerReset channel that this loop is
+				// the only reader of
```
The problem with that is that `l.getToken(ctx)` (may) mutate things in `l` without locking, so doing this in a separate goroutine can corrupt things if a second `<-l.refreshTimer.C` fires while the first goroutine is still running.

I think it would be better to just change the channel to be buffered?

```patch
--- a/pkg/client/connector/userd_auth/login.go
+++ b/pkg/client/connector/userd_auth/login.go
@@ -100,7 +100,10 @@ func NewLoginExecutor(
 		// in a running state with 1 minute left, and then immediately stop it below with
 		// resetRefreshTimerUnlocked.
 		refreshTimer:      time.NewTimer(1 * time.Minute),
-		refreshTimerReset: make(chan time.Duration),
+		// The "refresh" goroutine is the only reader of refreshTimerReset, and
+		// "refresh"->l.getToken()->l.tokenSource.Token()->l.tokenCB() may call
+		// l.resetRefreshTimer() to write to it, so we need a buffer size of at least 1.
+		refreshTimerReset: make(chan time.Duration, 1),
 	}
 	ret.oauth2ConfigMu.Lock()
 	ret.loginMu.Lock()
```

LukeShu reviewed (changes requested):
Good job tracking down the deadlock! But I'm not sold on the fix.

Commit:
Fix deadlock in login handler

The access-token "refresh" routine has a select case that calls the
`loginExecutor.getToken` method which in turn would retrieve the `Token`
and potentially end up in a callback writing to the `refreshTimerReset`
channel that was read by another case in that same select, which lead
to a dead-lock because there would be nothing reading that channel.

This commit makes the `refreshTimerReset` channel buffered so that
the deadlock is avoided.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

LukeShu reviewed (approved):
Hit it!
